internal/stores/objects.go

```diff
@@ -523,10 +539,45 @@ func deleteObject(tx *gorm.DB, key string) error {
 func (s *SQLStore) get(key string) (dbObject, error) {
 	var obj dbObject
 	tx := s.db.Where(&dbObject{ObjectID: key}).
-		Preload("Slabs.Slab.Shards.Contracts.Host").
+		Preload("Slabs.Slab.Shards.DBSector.Contracts.Host").
 		Take(&obj)
 	if errors.Is(tx.Error, gorm.ErrRecordNotFound) {
 		return dbObject{}, ErrObjectNotFound
 	}
 	return obj, nil
 }
+
+// TODO: Should we return slabs which are below MinShards? I think so because
+// the caller should handle that and then manually delete them if necessary.
+func (s *SQLStore) slabsForRepair(n int, failureCutoff time.Time) ([]uint, error) {
+	failureQuery := s.db.Model(&dbSlab{}).
+		Select("id as slab_id").
+		Where("last_failure < ?", failureCutoff.UTC())
+	inner := s.db.Table("(?)", failureQuery).
+		Select("slab_id, db_sector_id as sector_id").
+		Joins("LEFT JOIN shards ON slab_id = shards.db_slab_id")
+	middle := s.db.Table("(?)", inner).
+		Select("slab_id, sector_id, db_contract_id as contract_id").
+		Joins("LEFT JOIN contract_sectors ON sector_id = contract_sectors.db_sector_id")
+	outer := s.db.Table("(?)", middle).
+		Select("slab_id").
+		Joins("LEFT JOIN contracts ON contract_id = contracts.id").
+		Where("contract_id IS NULL OR is_good IS 0").
+		Group("slab_id").
+		Order("COUNT(slab_id) DESC").
+		Limit(n)
+
+	var slabIDs []uint
+	err := outer.Find(&slabIDs).Error
+	return slabIDs, err
+}
+
+// MarkSlabsFailure sets the last_failure field for the given slabs to the
+// current time.
+func (s *SQLStore) MarkSlabsFailure(slabIDs []uint) error {
```

Comment:
wondering whether we should assert the amount of updated slabs equals the amount of `slabIDs` - usually you have something like `modifiedCount` or whatever on the return value of your query

Reply:
Do you expect this to return an error if the updated count doesn't match? Or would it just log a warning?

Reply:
not sure, I just made the remark because `DELETE where ID in x` queries that don't report a modified count can be sneaky

Reply:
maybe I'm overthinking it, fine to leave as is for now, just triggered me when reading it initially

Reply:
I'll add a return value in the next PR touching this to return the number of modified sectors. Returning an error is a bit awkward since the caller's assumption would be that an error doesn't apply any change.